### PR TITLE
pint units (GEP 10): annotate METTSIM end to end (worked example)

### LINKED ISSUE #121: pint units: annotate mettsim + switch check ON + CI test over all dates
## What to build

Annotate the entire mettsim policy system with units, tag all bare literals in mixed-unit arithmetic, and switch the always-on fail_if from crafted fixtures to the real assembled environment. Add a CI test that runs the data-independent check (per-function + edge) across all mettsim policy dates. End-to-end proof that the framework holds on a complete system and the agnostic/multi-currency path works.

## Acceptance criteria

- [ ] every mettsim function/input/param carries a unit; nulls resolved
- [ ] all literals in mixed-unit arithmetic tagged
- [ ] always-on fail_if active on the assembled mettsim environment
- [ ] CI test runs the unit check over all mettsim dates and is green
- [ ] full ttsim + mettsim suite green

## Blocked by

- https://github.com/ttsim-dev/ttsim/issues/119
- https://github.com/ttsim-dev/ttsim/issues/120

### PR DESCRIPTION
**Pull request by MImmesberger**
Closes #121. Final PR of the GEP-10 stack (on top of #125 → #124 → #123 → #122).

Mettsim-focused by design: the check machinery lives in #124; this PR annotates mettsim, flips the check on, and adds the CI test.

## Switch ON

- Wire #124's environment-level checks into the interface DAG as two **always-on** `fail_if` nodes — `tt_units_are_missing` and `tt_units_are_inconsistent` — and add them to `MainTarget.fail_if`. Every assembled policy environment is now checked at build time.
- `testing_utils.check_env_units()` runs both checks data-independently over the full derivable environment at a policy date; `test_mettsim_units_are_complete_and_consistent` exercises it over **all mettsim policy dates** (GEP 10 Layer 1 as a CI test).
- ttsim's own test-fixture environments are annotated — mandatory units now apply to every environment the suite assembles.

## mettsim annotation (the worked example for gettsim #1192)

All policy inputs, ~23 policy functions, param functions, aggregations, and YAML parameters carry tokens. Highlights:

- `wealth_tax.tax_rate`: `SHARE_FLOW` + `reference_period: Year` → resolves `1/year`, so `wealth × tax_rate = CURRENCY/year` (the stock×rate pattern from the issue notes; concretely caught by the falsy-boolean dry-run if the time component is dropped).
- `subsistence_income_level`: a per-person amount — `CURRENCY` per `Month`; multiplying by a (dimensionless) head count is a plain scaling that stays a monthly flow.
- `child_tax_credit.schedule`: heterogeneous dict with **per-leaf tokens**; its consumer is concretely verified through the dict subscript.
- A SUM over a bool (`number_of_adults_fam`) is a head count → dimensionless (`unit=None`); lookup-table topups likewise; `bounty_per_orc` is `CURRENCY`.
- Boolean functions/inputs (`child`, `adult`, …) carry **no** annotation.

## The 2020 currency reform of Middle Earth (end-to-end EUR/DM proof)

Until 2020-01-01 the realm's unit of account is the Shire's **silver penny**; from the reform on, Gondor's **castar** (1 castar = 4 silver pennies).

- Every currency-denominated parameter carries a concrete token; pre-reform parameters are `SILVER_PENNY_*`.
- The wealth threshold and the standard payroll schedule carry **pure-redenomination changeover entries** (per-entry unit overrides, #125); the subsistence level's 2020 re-legislation is denominated in castar. A test asserts the changeover is value-continuous in either run currency.
- **Policy cases run in the period-appropriate currency**: pre-reform cases declare `currency: silver_penny` in their `info:` block (threaded through `testing_utils`) and keep their numbers; 2025 cases are re-expressed in castar (currency inputs ÷ 4, expectations regenerated — exactly old ÷ 4 modulo rounding).
- New end-to-end test: the same household run in castar and in silver pennies yields exactly fourfold currency outputs.
- Derived functions (time-conversion variants, aggregations) inherit the *agnostic* counterpart of a concrete source token — they compute on already-converted run-currency values.

## Drive-by

- **`net_income_parents_m` → `net_wealth_parents`**: the Python-3.14 bug reproducer computes `wealth * 0.8` (a stock); the `_m` flow suffix was dimensionally wrong and the new check caught it. Reproducer mechanics and expected values unchanged.

## Test plan

- [x] All CI green (tests on py311–py314 incl. jax across Linux/macOS/Windows; ty; codecov)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src_mettsim/mettsim/middle_earth/payroll_tax/child_tax_credit/annotation_bug_reproducer.py [hunk 1]
```diff
@@ -25,12 +25,14 @@ def child_in_household_for_bug_reproducer(p_id: int) -> bool:
...
-    leaf_name="net_income_parents_m",
+    leaf_name="net_wealth_parents",
     rounding_spec=RoundingSpec(
         base=1, direction="down", reference="Python 3.14 bug reproducer"
     ),
+    # Derived from `wealth`, so a stock: no time-unit suffix in the name.
+    unit="CURRENCY",
 )
-def net_income_parents_m(
+def net_wealth_parents(
```
Comment (MImmesberger) [inline src_mettsim/mettsim/middle_earth/payroll_tax/child_tax_credit/annotation_bug_reproducer.py:34]: `_m` implies `CURRENCY/month`, but `wealth` is a stock resolving to `CURRENCY`. Hence the rename

File: src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py [hunk 2]
```diff
@@ -31,6 +34,7 @@ def number_of_individuals_fam(fam_id: int) -> int:
 @policy_function(
     end_date="2019-12-31",
     leaf_name="requirement_fulfilled_fam",
+    unit=Unit.DIMENSIONLESS,
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py:35]: Cross-reference: this boolean-returning body is one of the functions the environment checker skips, so any dimensional comparison inside it is not checked — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673870).

File: src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py [hunk 3]
```diff
@@ -46,6 +50,7 @@ def requirement_fulfilled_fam_not_considering_children(
 @policy_function(
     start_date="2020-01-01",
     leaf_name="requirement_fulfilled_fam",
+    unit=Unit.DIMENSIONLESS,
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py:51]: Cross-reference: this boolean-returning body is one of the functions the environment checker skips, so any dimensional comparison inside it is not checked — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673870).

File: src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py [hunk 4]
```diff
@@ -58,31 +63,35 @@ def requirement_fulfilled_fam_considering_children(
...
+)
 def number_of_family_members_considered_fam(
     number_of_individuals_fam: int,
     max_number_of_family_members: int,
 ) -> int:
     return min(number_of_individuals_fam, max_number_of_family_members)
 
 
-@policy_function(vectorization_strategy="vectorize")
+@policy_function(vectorization_strategy="vectorize", unit=Unit.DIMENSIONLESS)
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py:86]: Cross-reference: this boolean-returning body is one of the functions the environment checker skips, so any dimensional comparison inside it is not checked — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673870).

File: src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py [hunk 5]
```diff
@@ -58,31 +63,35 @@ def requirement_fulfilled_fam_considering_children(
...
 def child(
     age: int,
     max_age_children: int,
 ) -> bool:
     return age <= max_age_children
 
 
-@policy_function(vectorization_strategy="vectorize")
+@policy_function(vectorization_strategy="vectorize", unit=Unit.DIMENSIONLESS)
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py:94]: Cross-reference: this boolean-returning body is one of the functions the environment checker skips, so any dimensional comparison inside it is not checked — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673870).

File: src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py [hunk 6]
```diff
@@ -58,31 +63,35 @@ def requirement_fulfilled_fam_considering_children(
...
 def adult(
     age: int,
     max_age_children: int,
 ) -> bool:
     return age > max_age_children
 
 
-@policy_function(vectorization_strategy="vectorize")
+@policy_function(vectorization_strategy="vectorize", unit=Unit.DIMENSIONLESS)
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/housing_benefits/eligibility/eligibility.py:102]: Cross-reference: this boolean-returning body is one of the functions the environment checker skips, so any dimensional comparison inside it is not checked — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673870).

File: src_mettsim/mettsim/middle_earth/wealth_tax/wealth_tax.py [hunk 7]
```diff
@@ -1,18 +1,18 @@
...
-    tax_rate: float,
+    tax_rate_y: float,
     exempt_from_wealth_tax: bool,
 ) -> float:
-    return 0.0 if exempt_from_wealth_tax else wealth * tax_rate
+    return 0.0 if exempt_from_wealth_tax else wealth * tax_rate_y
 
 
-@policy_function()
+@policy_function(unit=Unit.DIMENSIONLESS)
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/wealth_tax/wealth_tax.py:15]: Cross-reference: this boolean-returning body is one of the functions the environment checker skips, so any dimensional comparison inside it is not checked — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673870).

File: src_mettsim/mettsim/middle_earth/orc_hunting_bounty/orc_hunting_bounty.py [hunk 8]
```diff
@@ -22,7 +22,7 @@ class BountyPerOrc:
     large_orc: BountyPerLargeOrc
 
 
-@param_function()
+@param_function(unit=Unit.CURRENCY)
```
Comment (hmgaudecker) [inline src_mettsim/mettsim/middle_earth/orc_hunting_bounty/orc_hunting_bounty.py]: Cross-reference: `bounty_per_orc` returns an opaque `BountyPerOrc` without `verify_units=False`, which `_verify_one_body` neither checks nor rejects — see #124 (https://github.com/ttsim-dev/ttsim/pull/124#discussion_r3448673873).

**Review by hmgaudecker — COMMENTED** [review]
Cross-references to the GEP-10 audit findings, at the worked-example sites where they manifest.

File: src_mettsim/mettsim/middle_earth/orc_hunting_bounty/orc_hunting_bounty.py (hunk 8 above)
Comment (MImmesberger) [reply src_mettsim/mettsim/middle_earth/orc_hunting_bounty/orc_hunting_bounty.py]: `bounty_per_orc` now declares `verify_units=False` — with the M2 fix in #124, a body returning an opaque value (here the `BountyPerOrc` dataclass) is rejected unless it opts out. The boolean eligibility/wealth-tax bodies flagged in the sibling cross-refs are now dry-run (S2 fix) and pass. Pushed.

**Comment by MImmesberger** [comment]
Folded into #137 (combined GEP-10 foundation PR) to flatten the stacked-PR cascade. This PR's review history stays visible here; the branch is retained.
